x/concentrated-liquidity/internal/math/tick.go

```diff
@@ -1,52 +1,165 @@
+
+	// If the price is less than 1, we must reduce the increment distance by a factor of 10
+	if price.LT(sdk.OneDec()) {
+		kIncrementDistance = kIncrementDistance.Quo(sdkTenDec)
+	}
+
+	// Initialize the total price to 1, the current k to k_at_price_1, and the number of ticks passed to 0
+	totalPrice := sdk.OneDec()
+	ticksPassed := sdk.ZeroInt()
+	currentK := kAtPriceOne
+
+	// Set the currentIncrement to the kAtPriceOne
+	curIncrement := handleNegativeExponents(kAtPriceOne)
+
+	// Now, we loop through the k increments until we have passed the price
+	// Once we pass the price, we can determine what which k values we have filled in their entirety,
+	// as well as how many ticks that corresponds to
+	// In the opposite direction (price < 1), we do the same thing (just decrement k instead of incrementing).
+	// The only difference is we must reduce the increment distance by a factor of 10.
+	if price.GT(sdk.OneDec()) {
+		for totalPrice.LT(price) {
+			curIncrement = handleNegativeExponents(currentK)
+			maxPriceForCurrentIncrement := kIncrementDistance.Mul(curIncrement)
+			totalPrice = totalPrice.Add(maxPriceForCurrentIncrement)
+			currentK = currentK.Add(sdk.OneInt())
+			ticksPassed = ticksPassed.Add(kIncrementDistance.TruncateInt())
+		}
+	} else {
+		for totalPrice.GT(price) {
+			curIncrement = handleNegativeExponents(currentK)
+			maxPriceForCurrentIncrement := kIncrementDistance.Mul(curIncrement)
+			totalPrice = totalPrice.Sub(maxPriceForCurrentIncrement)
+			currentK = currentK.Sub(sdk.OneInt())
+			ticksPassed = ticksPassed.Sub(kIncrementDistance.TruncateInt())
+		}
+	}
+	// Determine how many ticks we have passed in the current k increment
+	ticksToBeFulfilledByCurrentK := price.Sub(totalPrice).Quo(curIncrement)
+
+	// Finally, add the ticks we have passed from the completed k values, as well as the ticks we have passed in the current k value
+	tickIndex := ticksPassed.Add(ticksToBeFulfilledByCurrentK.TruncateInt())
+
+	// Add a check to make sure that the tick index is within the allowed range
+	minTick, maxTick := GetMinAndMaxTicksFromK(kAtPriceOne, kIncrementDistance)
+	if tickIndex.LT(minTick) {
+		return sdk.Int{}, fmt.Errorf("tickIndex must be greater than or equal to %s", minTick)
+	}
+	if tickIndex.GT(maxTick) {
+		return sdk.Int{}, fmt.Errorf("tickIndex must be less than or equal to %s", maxTick)
+	}
+
+	return tickIndex, nil
+}
+
+// handleNegativeExponents treats negative exponents as 1/(10**|exponent|) instead of 10**-exponent
+// This is because the sdk.Dec.Power function does not support negative exponents
+func handleNegativeExponents(exponent sdk.Int) sdk.Dec {
+	if exponent.GTE(sdk.ZeroInt()) {
+		return sdkTenDec.Power(exponent.Uint64())
+	}
+	return sdk.OneDec().Quo(sdkTenDec.Power(exponent.Abs().Uint64()))
+}
+
+// GetMinAndMaxTicksFromK determines min and max ticks allowed for a given k value
+// This allows for a min spot price of 0.000000000000000001 and a max spot price of 200000000000 for every k value
```

Comment:
What does a spot price of `0.000000000000000001` for a given k value mean? I thought k was referring to precision? i.e. $k_{p1}$

Reply:
So imagine every k value allows a minimum of 0.000000000000000001 (through negative ticks) and a positive of 200000000000. The amount of power of 10s that must be cycled through would just change based on what the kp1 is

Reply:
1. No longer needed and is now removed
2. They can be pre-computed, but we would need to specify them for every allowed exponentAtPriceOne. I am instead utilizing this function:
https://github.com/osmosis-labs/osmosis/blob/8ac72ce10a33c10a9ab20b3e1db3e7902bbd363f/x/concentrated-liquidity/internal/math/tick.go#L168-L173
3. Max spot price is now 100000000000000000000000000000000000000, which is the same sig figs as gamm, but is a smaller number.

Reply:
Thanks!

Let's track number 2 in an issue. Also, could you please document 3. in `architecture.md` and explain how it relates to gamm max spot price

Reply:
https://github.com/osmosis-labs/osmosis/issues/3986

Reply:
Explanation of point three done here https://github.com/osmosis-labs/osmosis/pull/3874/commits/87a1d664c83024213eb2c7deb06ba0e748695213